### ISSUE TITLE
Angular transformer deadlocks on `_tests`

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#517]
After a few hours of debugging with @alorenzen, I've narrowed it down to here:

```dart
      await crawlImports(
          inputs,
          (id) => transform
              .readInputAsString(toBarbackAssetId(id))
              .catchError((_) => null));
```

... from `build_barback`. AFAICT, this was added by @natebosch for an internal customer that uses pub, but this same code causes circular dependencies in Angular itself to never succeed building when using with the `_tests` package:

```yaml
// Add this to _tests/pubspec.yaml at the bottom
transformers:
  - angular/transform/reflection_remover:
      $include: test/**_test.dart
  - angular/transform/codegen
  - test/pub_serve:
      $include: test/**_test.dart
```

I don't know if its safe to remove this, or make it a flag, or what the proper fix should be.

**Comment by natebosch** [comment]
If we remove this we're likely to end up hitting an issue later on.

Why is there a dependency from `angular` -> `_tests`? I wouldn't expect us to need any circular package level dependencies...

**Comment by matanlurey** [comment]
There is not a dependency, and I don't understand the details yet.

Only that all of my tests pass with that commented out, and they time out otherwise.

**Comment by natebosch** [comment]
While we were looking at the internal case I had some uncommitted changes that I was using to track down where deadlocks were coming from - @alorenzen might still have the patch. It basically boiled down to keeping a a `Map` from an asset to the `Set` of assets it's waiting on (I think before we did package level but we could modify this to file level) and then running the build, waiting for it to settle, then assuming whatever is still waiting has the deadlock.

I'd be very interested in finding whether we can narrow it down to that level.

I'm surprised the build succeed after skipping that step. In general I wouldn't expect it to be safe because a deadlock there should mean we'd hit a deadlock at a later step anyway, but I could be misremembering the details.

**Comment by natebosch** [comment]
Here is the high level reasoning behind that bit of code:

When we run the transformer with the `Resolver` we will end up needing to read transitive Dart imports at the moment we try to use the resolver. Grabbing a resolver adds a lock - only 1 Builder can be usign the Resolver at a time. Barback won't let us read any assets from packages which have not finished their Transformers yet so if we grab the Resolver lock, then can't do anything with it because one of the transitive imports is blocked by barback (and that package now can't get the resolver lock) we hit a deadlock.

This import crawling ensures the resolver is only "picked up" in dependency order so no resolver deadlock can happen. If we were to remove it then we'd be very likely to hit deadlocks in packages even where there is no cycle.

**Comment by matanlurey** [comment]
> While we were looking at the internal case I had some uncommitted changes that I was using to track down where deadlocks were coming from - @alorenzen might still have the patch. It basically boiled down to keeping a a `Map` from an asset to the `Set` of assets it's waiting on (I think before we did package level but we could modify this to file level) and then running the build, waiting for it to settle, then assuming whatever is still waiting has the deadlock.

I guess I'm confused. At least for the AngularDart template building step (`.template.dart`) files, we don't care if the `.template.dart` file has been built or not for our dependencies. We don't analyze it, and we don't rely on it in anyway. We _used_ to, but then we added the `.ng_placeholder` step. 

Is the placeholder code what enables us to remove this check?

**Comment by natebosch** [comment]
The limitation is that barback won't let you read _any_ file from another package until it is done with its Transformers. Even if we don't care about generated files we won't be able to read source files either.

**Comment by matanlurey** [comment]
So as long as I don't have any circular packages (and I'm totally fine with making that a restriction of AngularDart 4.0), it should be safe to disable your check @natebosch?

**Comment by alorenzen** [comment]
Even with the check, it is already a requirement that there are no circular packages.

**Comment by natebosch** [comment]
No, it will not be safe to disable the check. With this disabled the Resolver gets picked up in an arbitrary order - and if that order is not exactly dependency order you'll hit deadlock.

**Comment by matanlurey** [comment]
Then I guess I misunderstand why I need this check at all 😕

**Comment by natebosch** [comment]
This check exists to enforce that the build runs serially in dependency order, which is the only way the Resolver can avoid deadlock.

In `bazel` this is enforced by the build system. In `pub` we had to enforce it ourselves.

**Comment by matanlurey** [comment]
Do you have a suggestion to proceed? We _really_ need some tests enabled on travis in order to ensure `4.x` is success, but I'm blocked with deadlocks as of now _because_ of that check.

**Comment by matanlurey** [comment]
Like I'm totally fine if we need to add some sort of CLI-based checking, i.e.

```bash
$ pub run angular doctor

> Warning: You have a circular dependency. Bad bad bad!
```

**Comment by natebosch** [comment]
In the immediate term we need to figure out which resources are deadlocked here, and either fix the circular dependency (which we don't expect should exist 😞 ) or fix this code to work in that condition.

**Comment by natebosch** [comment]
Long term we want to move away from barback.

**Comment by alorenzen** [comment]
Wondering if this is going to be something specific to _tests. We might need to split compiler tests out from e2e tests, and only run the transformer on the later?

**Comment by matanlurey** [comment]
I guess the problem is even if I figure out where the problems are (I have a _rough_ idea with @alorenzen), there is no good way to prevent these from re-occurring (because we are internal first).

**Comment by matanlurey** [comment]
Ok @natebosch and @alorenzen and I figured this out - having the `reflection_remover` transformer run _before_ `codegen` causes a deadlock - `codegen` adds all crawled imports as a secondary input, which if this is after `reflection_remover` includes all generated files. That is a no-no!

**Comment by matanlurey** [comment]
A little stuck still.

If I remove the `reflection_remover` transformer from `angular_test` OR `_tests`, it works for the _other_ one. If they _both_ have `reflection_remover`, neither works.

**Comment by natebosch** [comment]
weird... Have you tried looking for the places that import `.template.dart` in the checked in code and avoiding those tests?

**Comment by matanlurey** [comment]
Yeah I thought we had some, we don't.

**Comment by matanlurey** [comment]
OK I confirmed if I add a way to "filter" the `crawlImports` from `build_barback`, everything works @natebosch. I'm going to see if I can isolate this inside of AngularDart for right now (my idea now is to write a Builder that wraps and filters out `.template.dart` files.

**Comment by matanlurey** [comment]
Another thing: the version of `build_barback` on pub is 0.2.x, the version internally is `0.3.0-dev` @natebosch. My assumption is that means... bad things?

**Comment by natebosch** [comment]
Yes we will need to get a new version published. I don't recall if there are blockers or if we just never got around to it.

**Comment by matanlurey** [comment]
Considering we use it internally, I imagine it's safe. I'm going to try and publish with @kevmoo today.

**Comment by matanlurey** [comment]
This looks fixed @ `build_barback 0.3.0`. Working...

**Comment by matanlurey** [comment]
Yup, `build_barback` looks good.
